### ISSUE TITLE
Improve build speed with incremental builds

### ISSUE THREAD

**Issue by xavez** [usebedrock/bedrock#281]
We might be able to leverage something like https://github.com/mpdroog/gulp-jade-recursive to improve build speed of larger projects. I think our build process overall is pretty speedy, but we might be able to speed it up more with this?

**Comment by Wolfr** [comment]
Do you mean to improve `gulp build` speed or render to browser / render page speed?

**Comment by Wolfr** [comment]
I checked the source code, very interesting and clever.

**Comment by Wolfr** [comment]
Did want to note that using gulp 4 will allow us to have more control about multiple tasks at the same time. I used gulp 4 in Bedrock Mini and I quite liked it. But at first sight it looks like we could possibly do both to optimize even further.

( https://github.com/usebedrock/bedrock-mini )

**Comment by Wolfr** [comment]
I looked further into this code for fun. I think there is some clever programming going on here using event streams etc. but not sure if we can use any of it.

The thing is that if we are making a full `dist` we need every file anyway, so it doesn't make sense to create a map of only what is affected, and in the use case of developing, only what is requested is rendered live using the express server, so we already have an optimal case.

What I found is that within Bedrock, the Pug template compilation is comprised of 3 discrete parts (docs, styleguide and actual templates).

Those parts seem to run quite a lot of functions to get the necessary metadata to work with (seemingly contained in `getDefaultLocals` function calls).

I think a lot of this can be cleaned up if we investigate it more but I am just not sure what some of the code really does.

It might be interesting to hear @thomastuts 's opinion on how to make his now 3 year old code faster ;).
